include/pybind11/detail/type_caster_base.h

```diff
@@ -470,41 +470,34 @@ PYBIND11_NOINLINE bool isinstance_generic(handle obj, const std::type_info &tp)
     return isinstance(obj, type);
 }
 
-PYBIND11_NOINLINE std::string error_string() {
-    if (!PyErr_Occurred()) {
-        PyErr_SetString(PyExc_RuntimeError, "Unknown internal error occurred");
-        return "Unknown internal error occurred";
+PYBIND11_NOINLINE std::string
+error_string(PyObject **exc_type, PyObject **exc_value, PyObject **exc_trace) {
+    if (*exc_type == nullptr) {
+        pybind11_fail(
+            "Internal error: pybind11::detail::error_string() called with exc_type == nullptr");
     }
 
-    error_scope scope; // Preserve error state
-
-    std::string errorString;
-    if (scope.type) {
-        errorString += handle(scope.type).attr("__name__").cast<std::string>();
-        errorString += ": ";
-    }
-    if (scope.value) {
-        errorString += (std::string) str(scope.value);
-    }
+    PyErr_NormalizeException(exc_type, exc_value, exc_trace);
```

Comment:
I am worried this will break a lot of downstream tests, and it's a relatively cheap operation and is noop if the PyErr is already normalized.

Reply:
That is only a deliberate experiment to see what happens.

Originally I got into this because I saw that error_string calls PyErr_SetString if no error is set.

Then I thought not adding the delayed m_lazy_string assignment in restore could help in two ways:
* Reduce the complexity, therefore avoiding potentially extremely harmful pitfalls of errors in the error handling.
* Be faster (the whole point of this PR).

I might have to peddle bacl, but I need to look/experiment more.

Reply:
> Unfortunately, PyErr_NormalizeException decref exc_value and exc_trace so this leaves the objects in a different state than they were before this PR.

I still need to drill down to learn how this leads to the test failures.

But to explain m current thinking:

Status quo is that `error_already_set` has the undocumented side effect*** of normalizing the Python exception. I see two "useful/good" options:

A: Document it, which means we're promising that it will stay that way.
B: Change the behavior to avoid the performance hit and document that.

I'm thinking A might be the way to go, but it also means that we're permanently accepting the performance hit.

Note that lazily building what() also leaves the object in a different state than before, potentially with similar consequences & surprises in the wild. It's a performance optimization with risks. Acceptable or not is a judgement call.

*** undocumented side effect:

Searching case-insensitive for "normali" on master:
```
include/pybind11/pytypes.h:    PyErr_NormalizeException(&exc, &val, &tb);
include/pybind11/pytypes.h:    PyErr_NormalizeException(&exc, &val2, &tb);
include/pybind11/numpy.h:        // Platform-dependent normalization
include/pybind11/detail/type_caster_base.h:    PyErr_NormalizeException(&scope.type, &scope.value, &scope.trace);
docs/advanced/cast/strings.rst:Normalizing combining characters before passing the character literal to C++
docs/advanced/cast/strings.rst:    >>> example.pass_wchar(unicodedata.normalize("NFC", combining_e_acute))
tests/test_numpy_array.cpp:    return {// Normalization
```